### ISSUE TITLE
TestHandleNodeSelector test flake

### ISSUE THREAD

**Issue by erictune** [GoogleCloudPlatform/kubernetes#10036]
In a PR that only changed docs, TestHandleNodeSelector failed on shippable.
https://app.shippable.com/builds/5582fe6f152d7e0b0098cd1a
says:
```
KUBE_RACE="-race" KUBE_COVER="y" KUBE_GOVERALLS_BIN="$HOME/gopath/bin/goveralls" KUBE_TIMEOUT='-timeout 300s' KUBE_COVERPROCS=8 KUBE_TEST_ETCD_PREFIXES="${KUBE_TEST_ETCD_PREFIXES}" KUBE_TEST_API_VERSIONS="${KUBE_TEST_API_VERSIONS}" ./hack/test-go.sh -- -p=2
Running tests for APIVersion: v1 with etcdPrefix: registry
...
W0618 17:31:57.275251 15439 docker.go:259] found a container with the "k8s" prefix, but too few fields (2): "k8s_unidentified"
I0618 17:31:57.275655 15439 container_gc.go:128] Removing unidentified dead container "/k8s_unidentified" with ID "2876"
I0618 17:31:57.278692 15439 image_manager.go:274] [ImageManager]: Removing image "image-0" to free 1024 bytes
I0618 17:31:57.279295 15439 image_manager.go:274] [ImageManager]: Removing image "image-0" to free 1024 bytes
I0618 17:31:57.279756 15439 image_manager.go:274] [ImageManager]: Removing image "image-0" to free 1024 bytes
I0618 17:31:57.280216 15439 image_manager.go:274] [ImageManager]: Removing image "image-0" to free 1024 bytes
I0618 17:31:57.281203 15439 image_manager.go:212] [ImageManager]: Disk usage on "" () is at 95% which is over the high threshold (90%). Trying to free 150 bytes
I0618 17:31:57.281342 15439 image_manager.go:274] [ImageManager]: Removing image "image-0" to free 450 bytes
I0618 17:31:57.281655 15439 image_manager.go:212] [ImageManager]: Disk usage on "" () is at 95% which is over the high threshold (90%). Trying to free 150 bytes
I0618 17:31:57.281744 15439 image_manager.go:274] [ImageManager]: Removing image "image-0" to free 50 bytes
W0618 17:31:57.307083 15439 kubelet.go:567] Data dir for pod "bothpod" exists in both old and new form, using new
W0618 17:31:57.309383 15439 kubelet.go:622] Data dir for pod "newpod", container "bothctr" exists in both old and new form, using new
I0618 17:31:57.350196 15439 plugins.go:56] Registering credential provider: .dockercfg
E0618 17:31:57.357026 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.423204 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.450256 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.450703 15439 kubelet.go:1835] error getting node: node not found
I0618 17:31:57.588034 15439 plugins.go:56] Registering credential provider: .dockercfg
E0618 17:31:57.589181 15439 kubelet.go:1739] Pod "_": HostPort is already allocated, ignoring: [[0].port: duplicate value '81/']
E0618 17:31:57.591178 15439 kubelet.go:1739] Pod "newpod_foo": HostPort is already allocated, ignoring: [[0].port: duplicate value '80/']
E0618 17:31:57.591468 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.594614 15439 kubelet.go:1835] error getting node: node not found
--- FAIL: TestHandleNodeSelector (0.00s)
kubelet_test.go:2318: status of pod "podA_foo" is not found in the status map
E0618 17:31:57.600623 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.602728 15439 kubelet.go:1739] Pod "pod2_": HostPort is already allocated, ignoring: [[0].port: duplicate value '80/']
E0618 17:31:57.603082 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.603282 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.666564 15439 kubelet.go:2121] Error updating node status, will retry: error getting node "": Node "" not found
E0618 17:31:57.666683 15439 kubelet.go:2121] Error updating node status, will retry: error getting node "": Node "" not found
E0618 17:31:57.666745 15439 kubelet.go:2121] Error updating node status, will retry: error getting node "": Node "" not found
E0618 17:31:57.666810 15439 kubelet.go:2121] Error updating node status, will retry: error getting node "": Node "" not found
E0618 17:31:57.666868 15439 kubelet.go:2121] Error updating node status, will retry: error getting node "": Node "" not found
E0618 17:31:57.690660 15439 kubelet.go:1365] Deleting mirror pod "foo_ns" because it is outdated
E0618 17:31:57.698968 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.714840 15439 kubelet.go:1835] error getting node: node not found
I0618 17:31:57.861009 15439 kubelet.go:872] Node was previously registered
W0618 17:31:57.861717 15439 kubelet.go:958] Port name conflicted, "fooContainer-foo" is defined more than once
W0618 17:31:57.861814 15439 kubelet.go:958] Port name conflicted, "fooContainer-TCP:80" is defined more than once
E0618 17:31:57.920199 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.923393 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.927939 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:57.932362 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:58.002995 15439 kubelet.go:1835] error getting node: node not found
E0618 17:31:58.025981 15439 kubelet.go:1835] error getting node: node not found
I0618 17:31:58.029290 15439 plugins.go:56] Registering credential provider: .dockercfg
I0618 17:31:58.095041 15439 plugins.go:56] Registering credential provider: .dockercfg
I0618 17:31:58.096872 15439 plugins.go:56] Registering credential provider: .dockercfg
I0618 17:31:58.158154 15439 plugins.go:56] Registering credential provider: .dockercfg
W0618 17:31:58.159426 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_foo_new_12345678_0"
W0618 17:31:58.159557 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_foo_new_12345678_0"
W0618 17:31:58.159651 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.159729 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.159908 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.160051 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.160146 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_foo_new_12345678_0"
W0618 17:31:58.160241 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_foo_new_12345678_0"
W0618 17:31:58.160879 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_bar_new_98765_0"
W0618 17:31:58.160981 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_bar_new_98765_0"
W0618 17:31:58.161063 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.161149 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.161322 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_bar_new_98765_0"
W0618 17:31:58.161447 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_bar_new_98765_0"
W0618 17:31:58.161542 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.161633 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.161879 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_bar_new_12345678_0"
W0618 17:31:58.161983 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_bar_new_12345678_0"
W0618 17:31:58.162078 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.162160 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.162298 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.162403 15439 docker.go:269] invalid container hash "hash123" in container "k8s_POD.hash123_foo_new_12345678_0"
W0618 17:31:58.162494 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_bar_new_12345678_0"
W0618 17:31:58.162576 15439 docker.go:269] invalid container hash "hash123" in container "k8s_bar.hash123_bar_new_12345678_0"
I0618 17:31:58.164822 15439 plugins.go:56] Registering credential provider: .dockercfg
E0618 17:31:58.165088 15439 kubelet.go:1835] error getting node: node not found
I0618 17:31:58.165618 15439 runonce.go:59] waiting for 1 pods
I0618 17:31:58.166095 15439 runonce.go:145] Container "bar" not running: api.ContainerState{Waiting:(*api.ContainerStateWaiting)(0xc20834c930), Running:(*api.ContainerStateRunning)(nil), Terminated:(*api.ContainerStateTerminated)(nil)}
I0618 17:31:58.166220 15439 runonce.go:111] pod "foo" containers not running: syncing
W0618 17:31:58.173560 15439 docker.go:259] found a container with the "k8s" prefix, but too few fields (5): "k8s_net_foo.new.test_abcdefgh_42"
I0618 17:31:58.173949 15439 runonce.go:124] pod "foo" containers synced, waiting for 1ms
W0618 17:31:58.178447 15439 docker.go:259] found a container with the "k8s" prefix, but too few fields (5): "k8s_net_foo.new.test_abcdefgh_42"
E0618 17:31:58.178664 15439 manager.go:742] Error examining the container: parse docker container name "/k8s_net_foo.new.test_abcdefgh_42" error: Docker container name "k8s_net_foo.new.test_abcdefgh_42" has less parts than expected [k8s net foo.new.test abcdefgh 42]
W0618 17:31:58.179478 15439 docker.go:259] found a container with the "k8s" prefix, but too few fields (5): "k8s_net_foo.new.test_abcdefgh_42"
I0618 17:31:58.179651 15439 runonce.go:107] pod "foo" containers running
I0618 17:31:58.179788 15439 runonce.go:72] started pod "foo"
I0618 17:31:58.179862 15439 runonce.go:81] 1 pods started
W0618 17:31:59.842085 15439 connection.go:126] Stream rejected: Unable to parse '' as a port: strconv.ParseUint: parsing "": invalid syntax
W0618 17:31:59.860739 15439 connection.go:126] Stream rejected: Unable to parse 'abc' as a port: strconv.ParseUint: parsing "abc": invalid syntax
W0618 17:31:59.967321 15439 connection.go:126] Stream rejected: Unable to parse '-1' as a port: strconv.ParseUint: parsing "-1": invalid syntax
W0618 17:32:00.026988 15439 connection.go:126] Stream rejected: Unable to parse '65536' as a port: strconv.ParseUint: parsing "65536": value out of range
W0618 17:32:00.104831 15439 connection.go:126] Stream rejected: Port '0' must be greater than 0
FAIL
coverage: 60.4% of statements
FAIL	github.com/GoogleCloudPlatform/kubernetes/pkg/kubelet	3.423s
```

**Comment by erictune** [comment]
@yujuhong might know about this test.

**Comment by erictune** [comment]
@jszczepkowski might also know about this test.

**Comment by yujuhong** [comment]
`git bisect` suggests that this was caused by #9728. I haven't read the PR yet. 
/cc @dchen1107 @justinsb

**Comment by dchen1107** [comment]
That is weird. How come shippable didn't complaint #9728 at the first place? But #9728 definitely is the biggest culprit here.

**Comment by erictune** [comment]
happened again:
https://app.shippable.com/builds/55830410152d7e0b0098d14a

seems pretty repeatable.

**Comment by justinsb** [comment]
I think Shippable did complain, but I think I dismissed it as a flake (and I can't see Shippable results because I'm not in the Google org).  A question I need to figure out is why _my_ e2e testing didn't fail.

**Comment by yujuhong** [comment]
@justinsb, the unit tests use a simple/test kubelet that does not go through the initialization path you wrote, so `nodeName` is always empty. This would only happen in the unit tests, and will always happen (not flaky). #10055 sets the `nodeName` for such test kubelets.

**Comment by dchen1107** [comment]
This is fixed by #10057. Thanks @yujuhong and @justinsb
